src/arkode/arkode_erkstep.c

```diff
@@ -847,6 +939,207 @@ int erkStep_TakeStep(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+                                           ark_mem->tn + ark_mem->h, &do_save);
+    if (do_save)
+    {
+      SUNAdjointCheckpointScheme_InsertVector(ark_mem->checkpoint_scheme,
+                                              ark_mem->checkpoint_step_idx,
+                                              step_mem->B->stages,
+                                              ark_mem->tn + ark_mem->h,
+                                              ark_mem->ycur);
+    }
+  }
+
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  erkStep_TakeStep_Adjoint:
+
+  This routine performs a single backwards step of the discrete
+  adjoint of the ERK method.
+
+  Since we are not doing error control during the adjoint integration,
+  the output variable dsmPtr should should be 0.
+
+  The input/output variable nflagPtr is used to gauge convergence
+  of any algebraic solvers within the step. In this case, it should
+  always be 0 since we do not do any algebraic solves.
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int erkStep_TakeStep_Adjoint(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+{
+  int retval = ARK_SUCCESS;
+
+  ARKodeERKStepMem step_mem;
+
+  /* access ARKodeERKStepMem structure */
+  retval = erkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  SUNLogDebug(ARK_LOGGER, "ARKODE::erkStep_TakeStep_ERK_Adjoint", "start-step",
+              "step = %li, h = %" SUN_FORMAT_G ", dsm = %" SUN_FORMAT_G
+              ", nflag = %d",
+              ark_mem->nst, ark_mem->h, *dsmPtr, *nflagPtr);
+
+  /* local shortcuts for readability */
+  SUNAdjointStepper adj_stepper = (SUNAdjointStepper)ark_mem->user_data;
+  sunrealtype* cvals            = step_mem->cvals;
+  N_Vector* Xvecs               = step_mem->Xvecs;
+  N_Vector sens_np1             = ark_mem->yn;
+  N_Vector sens_n               = ark_mem->ycur;
+  N_Vector sens_tmp             = ark_mem->tempv2;
+  N_Vector Lambda_tmp           = N_VGetSubvector_ManyVector(sens_tmp, 0);
+  N_Vector lambda_np1           = N_VGetSubvector_ManyVector(sens_np1, 0);
+  N_Vector* stage_values        = step_mem->F;
+
+  /* which adjoint step is being processed */
+  ark_mem->adj_step_idx = adj_stepper->final_step_idx - ark_mem->nst;
+
+  /* determine if method has fsal property */
+  sunbooleantype fsal = (SUNRabs(step_mem->B->A[0][0]) <= TINY) &&
+                        ARKodeButcherTable_IsStifflyAccurate(step_mem->B);
+
+  /* Loop over stages */
+  if (fsal) { N_VConst(SUN_RCONST(0.0), stage_values[step_mem->stages - 1]); }
+  for (int is = step_mem->stages - (fsal ? 2 : 1); is >= 0; --is)
```

Comment:
Same suggestion as in ARKStep
```suggestion
  /* For FSAL ERK methods, A[s-1][s-1] == b[s-1] = 0 so F[s-1] is always zero *
  if (fsal) { N_VConst(SUN_RCONST(0.0), stage_values[step_mem->stages - 1]); }

  /* Loop over stages */
  for (int is = step_mem->stages - (fsal ? 2 : 1); is >= 0; --is)
```